### PR TITLE
Fix: Hitting Enter key in a list item doesn't create a new list item [NO CHANGING SCHEMA]

### LINKED ISSUE #3474: bug: lists not working properly in the markdown widget
**Describe the bug**
Trying Gatsby + Netlify CMS Starter and it seems to work great except for the odd behavior of the Rich Text editor.  It doesn't seem to work correctly with lists.  It will start a list (1. ) but then when hitting enter it will just line the cursor underneath the first character of the previous list item with no number. And if I hit the numbered list item again it will indent a sub-list and then I have to shift-tab to move it back.  non ordered lists have the same issue.

**To Reproduce**
- Begin with the gatsby-starter-netlify-cms template
- Follow Getting Started (Without Netlify) - deploying to Cloud Foundry
- Configured and deployed the Netlify CMS GitHub Provider, GitHub repository and CMS plugin configuration for authentication
- Go to /admin and edit a new blog entry with a list.  First item begins with 1., but following items don't have a number but indented inline with first entry.  Works if edit as Markdown instead of Rich Text

**Expected behavior**
Expect the list item to generate a new list and each successive "Enter" button the list index should increment.

**Screenshots**

**Applicable Versions:**

 - Netlify CMS version: 2.11.32
 - Git provider: GitHub
 - OS: Windows 10
 - Browser version: Chrome 80.0.3987.122
 - Node.JS version: 12.14.1

**CMS configuration**
```yaml
backend:
  name: github
  base_url: https://fdic-ea-auth-d.app.cloud.gov
  repo: amaline/ea-test-site
  branch: master
  commit_messages:
    create: 'Create {{collection}} “{{slug}}”'
    update: 'Update {{collection}} “{{slug}}”'
    delete: 'Delete {{collection}} “{{slug}}”'
    uploadMedia: '[skip ci] Upload “{{path}}”'
    deleteMedia: '[skip ci] Delete “{{path}}”'

media_folder: static/img
public_folder: /img

collections:
  - name: "blog"
    label: "Blog"
    folder: "src/pages/blog"
    create: true
    slug: "{{year}}-{{month}}-{{day}}-{{slug}}"
    fields:
      - {label: "Template Key", name: "templateKey", widget: "hidden", default: "blog-post"}
      - {label: "Title", name: "title", widget: "string"}
      - {label: "Publish Date", name: "date", widget: "datetime"}
      - {label: "Description", name: "description", widget: "text"}
      - {label: "Featured Post", name: "featuredpost", widget: "boolean"}
      - {label: "Featured Image", name: "featuredimage", widget: image}
      - {label: "Body", name: "body", widget: "markdown"}
      - {label: "Tags", name: "tags", widget: "list"}

  - name: "pages"
    label: "Pages"
    files:
      - file: "src/pages/index.md"
        label: "Landing Page"
        name: "index"
        fields:
          - {label: "Template Key", name: "templateKey", widget: "hidden", default: "index-page"}
          - {label: Title, name: title, widget: string}
          - {label: Image, name: image, widget: image}
          - {label: Heading, name: heading, widget: string}
          - {label: Subheading, name: subheading, widget: string}
          - {label: Mainpitch, name: mainpitch, widget: object, fields: [{label: Title, name: title, widget: string}, {label: Description, name: description, widget: text}]}
          - {label: Description, name: description, widget: string}
          - {label: Intro, name: intro, widget: object, fields: [{label: Heading, name: heading, widget: string}, {label: Description, name: description, widget: text}, {label: Blurbs, name: blurbs, widget: list, fields: [{label: Image, name: image, widget: image}, {label: Text, name: text, widget: text}]}]}
          - {label: Main, name: main, widget: object, fields: [{label: Heading, name: heading, widget: string}, {label: Description, name: description, widget: text}, {label: Image1, name: image1, widget: object, fields: [{label: Image, name: image, widget: image}, {label: Alt, name: alt, widget: string}]}, {label: Image2, name: image2, widget: object, fields: [{label: Image, name: image, widget: image}, {label: Alt, name: alt, widget: string}]}, {label: Image3, name: image3, widget: object, fields: [{label: Image, name: image, widget: image}, {label: Alt, name: alt, widget: string}]}]}
      - file: "src/pages/about/index.md"
        label: "About"
        name: "about"
        fields:
          - {label: "Template Key", name: "templateKey", widget: "hidden", default: "about-page"}
          - {label: "Title", name: "title", widget: "string"}
          - {label: "Body", name: "body", widget: "markdown"}
      - file: "src/pages/products/index.md"
        label: "Products Page"
        name: "products"
        fields:
          - {label: "Template Key", name: "templateKey", widget: "hidden", default: "product-page"}
          - {label: Title, name: title, widget: string}
          - {label: Image, name: image, widget: image}
          - {label: Heading, name: heading, widget: string}
          - {label: Description, name: description, widget: string}
          - {label: Intro, name: intro, widget: object, fields: [{label: Heading, name: heading, widget: string}, {label: Description, name: description, widget: text}, {label: Blurbs, name: blurbs, widget: list, fields: [{label: Image, name: image, widget: image}, {label: Text, name: text, widget: text}]}]}
          - {label: Main, name: main, widget: object, fields: [{label: Heading, name: heading, widget: string}, {label: Description, name: description, widget: text}, {label: Image1, name: image1, widget: object, fields: [{label: Image, name: image, widget: image}, {label: Alt, name: alt, widget: string}]}, {label: Image2, name: image2, widget: object, fields: [{label: Image, name: image, widget: image}, {label: Alt, name: alt, widget: string}]}, {label: Image3, name: image3, widget: object, fields: [{label: Image, name: image, widget: image}, {label: Alt, name: alt, widget: string}]}]}
          - {label: Testimonials, name: testimonials, widget: list, fields: [{label: Quote, name: quote, widget: string}, {label: Author, name: author, widget: string}]}
          - {label: Full_image, name: full_image, widget: image}
          - {label: Pricing, name: pricing, widget: object, fields: [{label: Heading, name: heading, widget: string}, {label: Description, name: description, widget: string}, {label: Plans, name: plans, widget: list, fields: [{label: Plan, name: plan, widget: string}, {label: Price, name: price, widget: string}, {label: Description, name: description, widget: string}, {label: Items, name: items, widget: list}]}]}
```

### PR DESCRIPTION
**Pull request by bytrangle**
fixes #3474 

## Summary
A different approach to the above issue without changing the underlying schema.

## To-dos

- [x] Create a new list item below when hitting Enter key on a list item.
- [x] Fix tests for toolbar button
- [x] Fix tests for Enter key
- [x] Insert a new list item at the beginning of a nested list when pressing Enter key on parent list
- [x] Modify Enter key handler to account for case when the list item is not the last child in its parent list
- [x] Fix Enter key on nested list item
- [x] Fix tests for Backspace key
- [x] Fix tests for Tab key

### Setting up
1. After running the app, go to http://localhost:8080.
2. Click on any of the pre-created blog post.
3. Toggle the text editor to rich text.
4. Place cursor at the end of the last line in the text editor and press Enter or Return key to create a new empty paragraph block.
5. Type anything in the new block.

### Toggling list block
1. With the cursor in any block you want, click the Bulleted List button to turn a block into a bulleted list.
2. Move cursor to the end of the bulleted list item block and hit Enter.
3. *Expect*: A new list item underneath, marked by a bullet icon.
4. Inspect the element in the Developer Tools. *Expect*: Under the block that receives the Enter key is a new `<li> element containing a `<p>` element that wraps a text node .

insert-list-item-with-old-schema

### CODE REVIEW

**Comment by bytrangle** [comment]
Thanks for the reformat @erezrokah . What script did you run? I've tried running `lint`, `lint:js`, `lint:format` but it didn't auto-add the semicolon and spaces between curly brackets.

**Comment by erezrokah** [comment]
I ran [`yarn format` ](https://github.com/netlify/netlify-cms/blob/master/CONTRIBUTING.md#format).

Please let me know if it doesn't work for you (I think we had another contributor with a similar issue).

**Comment by bytrangle** [comment]
Luckily, it works for me. Thank you. 👍

**Comment by bytrangle** [comment]
Hello @erezrokah. I've noticed that my fix is failing a bunch of Cypress tests 😱.
I'm going to look at it tomorrow, but I'm not familiar with Cypress. What command should I run for e2e testing? I ran `yarn test:e2e` but it took an awfully long time.

**Comment by erezrokah** [comment]
> I ran `yarn test:e2e:dev` and it seems to be what I'm looking for.

Hi @bytrangle, thank you for looking into it (I meant to do it too after merging some of the smaller PRs that are currently open).

That commands works, see also https://github.com/netlify/netlify-cms/tree/master/cypress#debug-tests

**Comment by bytrangle** [comment]
> You should also be able to download the tests video recordings from the [GitHub action run](https://github.com/netlify/netlify-cms/pull/5550/checks?check_run_id=2942030388)

Cool, I'm not aware that there're videos inside the Github action.

So, I know why I failed the first test for the list plugins:
 **“creates nested list when selection is collapsed in non-first block of list item”**.

Here's how you produce it:
1. Click Unordered List button.
2. Type “foo”.
3. Hit Enter key.
4. Click unordered list button again.

The fourth step signals that you want to create a nested list inside the current list.

The content after step 3:
**Before Enter key handler change**
```
<ul>
  <li>
    <p>foo</p>
    <p></p> <!-- The focus and startBlock is here -->
</ul>
```
new-paragraph-block-on-enter

**After Enter key handler change**
```
<ul>
  <li>
    <p>foo</p>
  <li>
    <p></p> <!-- The focus and startBlock is here -->
  </li>
</ul>
```
new-list-item-block-on-enter

Next, when you click Unordered List button on the new paragraph block, it will call the `toggleList` with the relevant code here:
```
// ...
case 'list-item': {
  const listItem = target;
  const list = editor.value.document.getParent(listItem.key);
  if (!editor.isFirstChild(startBlock)) { // WHAT WILL GET CALLED BEFORE CHANGES TO ENTER KEY HANDLER
    editor.wrapInList(type);
  } else if (list.type !== type) {
    editor.toggleListItemType(listItem);
  } else {
    editor.unwrapListItem(listItem); // WHAT WILL GET CALLED AFTER CHANGES TO ENTER KEY HANDLER
  }
    break;
}
```
**Before change**: `startBlock` is the second child, so the first condition `!editor.isFirstChild(startBlock)` is satisfied and it will turn the paragraph block into a nested list.
**New change**: `startBlock` will always be the first child of its parent, the list item block so this condition is skipped until we meet the `else` statement, which will unwrap the list item and turn it into a default paragraph block not belonging to a list block.

The content after step 4:
**Before change**
```
<ul>
  <li>
    <p>foo</p>
    <ul>
      <li>
        <p></p>
      </li>
    </ul>
  </li>
</ul>
```
This  is also what the test expects.

**After change**
```
<ul>
  <li>
    <p>foo</p>
  </li>
</ul>
<p></p>
```
This is what the test receives, so unsurprisingly it failed.

I suggest _rewriting the output that the test expects to receive_  for three reasons:
*This output is in line with the behaviour of many rich text editor (Google Docs and Dropbox Paper, to name a few).*

A button in the control bar of a rich text editor toggles a style on and off, just like a dark mode switcher. A block is either an unordered list block, or it is not. There is no third scenario.

When you are on an empty list item block of an unordered list and you click the Unordered List button, that list item will be converted to a standalone paragraph block.

At this point, the Unordered List button is highlighted because we are inside an unordered list block, so it may not occur to many people that clicking it again will create a nested list block.

*There is already a Tab key handler that will take care of making a nested list.*
This is also a common way to create nested list in many rich text editors.

*We can add a new button for creating a nested list block*.
This is a design feature that I learn from Typora. I also think that it is more intuitive than attaching the list indentation method to the current list buttons.

What do you think?

**Comment by erezrokah** [comment]
Woo, @bytrangle thank you for the detailed explanation - this all makes sense.
I think you're correct and we should update the test case.

We should either update the output (and then the name of the test) or fix the test to create a nested list via the `tab` key.

In general we should cover the basic use cases:
1. Clicking the list button creates a list.
2. Clicking the list button, entering some text and hitting `enter` creates a new list item.
3. Clicking the list button, entering some text and hitting `enter` creates a new list item. Clicking the list button again changes the current list item to a paragraph.
4. Clicking the list button, entering some text and hitting `enter` creates a new list item. Hitting `tab` creates a nested list.

As for a dedicated button for nested lists, we can discuss this in a new issue.

**Comment by bytrangle** [comment]
I like those use cases.
But there are more edge cases to come in the test and I need your guidance 😅..
First example, corresponding to the test **“converts empty nested list item to empty block in parent list item”** for toolbar buttons:

Click List button. Type 'foo'. Hit Enter to create a new list item. Hit Tab to turn this list item into a nested list. Click the List button again.
```
<ul>
  <li>
    <p>foo</p>
    <ul>
      <li>
        <p></p> <!-- CLICK LIST BUTTON HERE -->
      </li>
    </ul>
</ul>
```

What do you think should happen? In most editors, this would turn the empty nested list back into a paragraph block.

But where should this paragraph block belong? Sibling of the *foo* block or sibling of the outer `<ul>`?

I'm going to ask about other cases later because I'm getting lost in all those scenarios!.

**Comment by erezrokah** [comment]
> I'm going to ask about other cases later because I'm getting lost in all those scenarios!!

To me it doesn't seem you're getting lost - it seems you're perfectly understanding the existing behavior and why we should change it.

 
> But where should this paragraph block belong? Sibling of the _foo_ block or sibling of the outer `<ul>` at the top level?

As a sibling of `foo` make more sense to me

**Comment by bytrangle** [comment]
Hello @erezrokah  What about this case:
- Create a list block with three items: foo, bar and baz.
- Select *bar* and *baz* only.
- Click List button. They will be converted to paragraph block.
- Click List button again. Do you want *bar* and *baz* to be
  - two sibling separate list items like in the beginning? OR
  - one list items with two paragraph blocks, one contains *bar* and the other contains *baz*
I'm leaning towards the first approach as it is a more common behaviour in rich text editors..

**Comment by bytrangle** [comment]
And this case:
- Clicking the list button, type *foo*.
- Hit Enter to create a new list item. Hitting tab to creates a nested list. Type “bar”.
- Hit Enter to create a new list item inside the nested list. Hit Enter again.

What should happen?
Does the empty list item become a sibling of the list item that contains *foo*, like this:
```
<ul>
  <li>
    <p>foo</p>
    <ul>
        <li>
          <p>bar</p>
        </li>
    </ul>
  </li>
  <li>
    <p></p>
  </li>
</ul>
```
Or it becomes a paragraph and a sibling of the nested list that contains *bar*

```
<ul>
  <li>
    <p>foo</p>
    <ul>
      <li>
        <p>bar</p>
      </li>
    </ul>
    <p></p>
  </li>
</ul>
```

There are a few tests that require us to be explicit about this.

**Comment by erezrokah** [comment]
I'm loving your attention to details working on this issue 🚀 

> Hello @erezrokah What about this case:
> 
> * Create a list block with three items: foo, bar and baz.
> * Select _bar_ and _baz_ only.
> * Click List button. They will be converted to paragraph block.
> * Click List button again. Do you want _bar_ and _baz_ to be
>   
>   * two sibling separate list items like in the beginning? OR
>   * one list items with two paragraph blocks, one contains _bar_ and the other contains _baz_
>     I'm leaning towards the first approach as it is a more common behaviour in rich text editors..

I agree the first one makes more sense (I think the expectation is that clicking the same button again will undo the previous operation in this case).


> And this case:
> 
> * Clicking the list button, type _foo_.
> * Hit Enter to create a new list item. Hitting tab to creates a nested list. Type “bar”.
> * Hit Enter to create a new list item inside the nested list. Hit Enter again.

I'd expect hitting enter on an empty list item would go one level up (the same if you create `foo`, hit enter, then hit enter again, it will go up to the top level).


> Or it becomes a paragraph and a sibling of the nested list that contains _bar_

If I understand you correctly, this is the expected result, also this photo (after the final enter key):
image

**Comment by bytrangle** [comment]
As for the first scenario, the second option is what will happen with Netlify CMS editor. I'm going to fix it in another request if you agree. Otherwise, I'm running in this circle: just attempt to fix test --> discover an unexpected behaviour --> fix again --> previous test fixes fail😅..

**Comment by erezrokah** [comment]
> As for the first scenario, the second option is what will happen with Netlify CMS editor. I'm going to fix it in another request if you agree. Otherwise, I'm running in this circle: just attempt to fix test --> discover an unexpected behaviour --> fix again --> previous test fixes fail😅..

I'm good with breaking these fixes into separate PRs 👍

**Comment by bytrangle** [comment]
The latest commits cause the build to fail. Is it because I didn't run lint before pushing? If so, am I supposed to push a new commit with the codes already linted?

**Comment by bytrangle** [comment]
> 
> I'm good with breaking these fixes into separate PRs 👍

What is your recommended way to split this PR? I've never done this, but what I'm thinking is:
- git branch fix-e2e-tests-for-list-plugin 
- in the current branch, rebase to remove those commits related to fixing tests.

I've also added a TO-DOS section to this PR's description. Please tell me what part you want to keep in this PR, and what should go to a new PR.

**Comment by erezrokah** [comment]
> The latest commits cause the build to fail. Is it because I didn't run lint before pushing? If so, am I supposed to push a new commit with the codes already linted?

Yes 💯 

> What is your recommended way to split this PR? I've never done this, but what I'm thinking is:

We'll need to get the tests passing for this PR so I would:
1. Keep the current fix described in the title of this PR
2. Fix the relevant tests to match the fix
3. If any other tests are failing, update those to reflect the new behavior after the fix, as long as the new behavior is not making things worse.
4. Open a new issue(s) to fix other use cases

Let me know if that makes sense

**Comment by bytrangle** [comment]
If I understand correctly, I should continue to fix all the failing tests for the List widget because all of them involves using the Enter key, which has been modified and is the main reason why the tests fail.

I haven't examined the tests for Quote widget in detail, but the failing ones also clicks the Unordered List button, and hit Enter key, which naturally would cause them to fail. 

Do I get it right?

**Comment by bytrangle** [comment]
There is a failing test for the Backspace that I'm not sure how to fix:
1. Click Unordered List button
2. Type “foo”
3. Enter
4. Hit Backspace

The output is supposed to be:
```
<ul>
  <li>
    <p>foo</p>
    <p></p>
  </li>
</ul>
```

But this output is exactly the same as what is expected in the next failed test: “Removes the list item if list not empty”.

I was thinking of removing this test, but at the same time I think the behaviour of the Backspace should be fixed.

If you're on an empty list item block, and you hit backspace, most editors would turn that block into a paragraph block.

turn-empty-list-item-into-paragraph-backspace

So I think I should comment out this test “_removes empty block in non-empty list item_” so that it won't fail, and I would remember it as an issue to be fixed next.

Lemme know if that makes sense.

**Comment by erezrokah** [comment]
Hi @bytrangle, I'm not sure where we stand with this after the recent commits.
Based on the latest commit from this branch, looks like we should remove `removes empty block in non-empty list item`.

**Comment by erezrokah** [comment]
> If I understand correctly, I should continue to fix all the failing tests for the List widget because all of them involves using the Enter key, which has been modified and is the main reason why the tests fail.
> 
> I haven't examined the tests for Quote widget in detail, but the failing ones also clicks the Unordered List button, and hit Enter key, which naturally would cause them to fail.
> 
> Do I get it right?

Sorry I missed this comment. You got it right

**Comment by bytrangle** [comment]
> Hi @bytrangle, I'm not sure where we stand with this after the recent commits.
> Based on the latest commit from this branch, looks like we should remove `removes empty block in non-empty list item`.

Yes, that's also what I'm thinking.

**Comment by erezrokah** [comment]
Hi @bytrangle, is this ready to review? Looks like the tests are passing

**Comment by bytrangle** [comment]
Yes, please go ahead. Is there anything else I need to do?

**Comment by erezrokah** [comment]
> Is there anything else I need to do?

Not at the moment. I'll review and comment if needed 👍

File: packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/List.js [hunk 1]
```diff
@@ -268,9 +268,46 @@ function ListPlugin({ defaultType, unorderedListType, orderedListType }) {
 
         if (listOrListItem.type === 'list-item') {
           const listItem = listOrListItem;
-
+          console.log({ listItem });
```
Comment (erezrokah) [inline packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/List.js:271]:
```suggestion
```

File: packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/BreakToDefaultBlock.js [hunk 2]
```diff
@@ -12,6 +12,11 @@ function BreakToDefaultBlock({ defaultType }) {
         editor.delete();
         return next();
       }
+      /** If the block that receives the Enter key is a child of a list item block, we'll
+       * skip this plugin and create a new list item block underneath instead.
+       */
+      const parent = document.getParent(startBlock.key);
+      if (parent.type === 'list-item') return next();
```
Comment (erezrokah) [inline packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/BreakToDefaultBlock.js:19]: Can you help me understand when this code it triggered? I couldn't find a scenario where I reach it

File: packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/List.js [hunk 3]
```diff
@@ -268,9 +268,46 @@ function ListPlugin({ defaultType, unorderedListType, orderedListType }) {
...
+             *      <li>
+             *        <p></p> ----- WHERE THE ENTER KEY HAPPENS
+             *      </li>
+             *    </ul>
+             *  </li>
+             * </ul>
+             */
+            const closestAncestorList = document.getClosest(
+              parentList.key,
+              block => block.type === (LIST_TYPES[0] || LIST_TYPES[1]),
```
Comment (erezrokah) [inline packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/List.js:298]:
```suggestion
              block => LIST_TYPES.includes(block.type),
```

I'm not sure if the original code was supposed to be `block.type === LIST_TYPES[0] || block.type === LIST_TYPES[1]`

File: packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/List.js [hunk 4]
```diff
@@ -285,7 +322,25 @@ function ListPlugin({ defaultType, unorderedListType, orderedListType }) {
...
-
+          const list = document.getParent(listItem.key);
+          if (LIST_TYPES.includes(list.type)) {
+            const newListItem = Block.create({
+              type: 'list-item',
+              nodes: [Block.create('paragraph')],
+            });
+            // Check if the targeted list item contains a nested list. If it does, insert a new list item in the beginning of that nested list.
+            const nestedList = listItem.findDescendant(
+              block => block.type === LIST_TYPES[0] || block.type === LIST_TYPES[1],
```
Comment (erezrokah) [inline packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/List.js:333]:
```suggestion
              block => LIST_TYPES.includes(block.type),
```

File: packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/List.js [hunk 5]
```diff
@@ -268,9 +268,46 @@ function ListPlugin({ defaultType, unorderedListType, orderedListType }) {
...
+             *    </ul>
+             *  </li>
+             * </ul>
+             */
+            const closestAncestorList = document.getClosest(
+              parentList.key,
+              block => block.type === (LIST_TYPES[0] || LIST_TYPES[1]),
+            );
+            if (closestAncestorList) {
+              const closestAncestorListItem = document.getParent(parentList.key);
```
Comment (erezrokah) [inline packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/List.js:301]: Can you share a scenario to reach this code? I wasn't able to reach here

**Review by erezrokah — CHANGES_REQUESTED** [review]
Woo, @bytrangle this is some amazing work 🚀 
I really appreciate the code and tests comments.

I left some notes - some are minor changes to the code, the others changes I didn't understand.

Please let me know what you think

**Comment by bytrangle** [comment]
@erezrokah Thank you for your review. I'm in the middle of something else, and I've forgotten some parts of it, so please let me get back to you in a few days.

File: packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/BreakToDefaultBlock.js (hunk 2 above)
Comment (bytrangle) [reply packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/BreakToDefaultBlock.js:19]: Oh, this code is a remnant from another (scraped) approach where I attempted to make list item blocks contain only text nodes, not pargraph block, like this:
```
<li data-slate-object="block" data-key="143">
  <span data-slate-object="text" data-key="144">
    <span data-slate-leaf="true" data-offset-key="144:0">
      <span data-slate-string="true">Try it out for yourself!</span>
    </span>
  </span>
</li>
```

With that schema, when you hit Enter on a list item, this condition inside the `BreakToDefaultBlock` plugin will be met:

```
if (
        selection.start.isAtEndOfNode(startBlock) &&
        startBlock.type !== "paragraph"
      ) {
        return editor.insertBlock("paragraph");
      }
```
because this plugin is loaded after the List plugin. This means instead of creating a new list item, we'll get another paragraph block on the level as the previous list item.

If you're interested, check out this [demo](https://codesandbox.io/s/slate-editor-v047-ptjv8?file=/src/components/VisualEditor.js) I made a month ago. Place cursor on the final line and click Bulleted List button. Then hit Enter.

But long story short, we no longer need this 😅.

**Comment by erezrokah** [comment]
> @erezrokah Thank you for your review. I'm in the middle of something else, and I've forgotten some parts of it, so please let me get back to you in a few days.

Sure no problem. Thank you for the quick response

File: packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/BreakToDefaultBlock.js (hunk 2 above)
Comment (bytrangle) [reply packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/BreakToDefaultBlock.js:19]: What do you suggest I do to this code? Remove the check for `list-item` and make a new commit, or rewrite the history?

Comment (erezrokah) [reply packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/BreakToDefaultBlock.js:19]: You can remove the code in a new commit. We squash merge PRs anyway so all the commits from this PR will show up as a single commit on the default branch.

File: packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/List.js (hunk 5 above)
Comment (bytrangle) [reply packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/List.js:301]: First, let's comment out the whole `if (closestAncestorList)` block.
In the editor:
- Click Unorder List button, type “foo”. Hit Enter.
- Type “bar”. Hit Enter.
- Hit Tab to indent the list. Type “baz”. Hit Enter.

Expected:
```
<ul>
  <li>
    <p>foo</p></li>
  <li>
    <p>bar</p>
    <ul>
      <li>
        <p>baz</p>
      </li>
      <li>
        <p></p>
      </li>
    </ul>
  </li>
</ul>
```
Making sure that the cursor is inside the empty new list item below “baz”, hit Enter again.

This is when the condition `editor.atStartOf(listItem)` will be met. The original code unwraps the list item and turns it back into a paragraph block on the same level with the nested list that contains “baz”.

new-paragraph-under-nested-list

Now, hit Backspace to delete the paragraph and bring the cursor back to the end of “baz”.

Bring back the `closestAncestorList` we've commented out.

A list item always has a parent of type `bulleted-list` or `numbered-list`, but does this parent have another list block above it? The code will check that. If it is, we'll move the current list item outside to become a direct child of that list block from above.

But I can see why you're confused. My code will result in false negative for `closestAncestorList`. So let's fix it like this:

```
const closestAncestorList = document.getClosest(
  parentList.key,
  block => LIST_TYPES.includes(block.type),
);
```
Back to the editor. With the cursor at the end of “baz”, hit Enter twice. This is what should happen:
 
move-nested-list-one-level-up

Does that make sense?

Comment (erezrokah) [reply packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/List.js:301]: > Does that make sense?

Yes it does. The purpose is to create a new list item at the end of the parent list instead of a paragraph, when hitting enter of an empty sub list item for a non empty list.

**Review by erezrokah — APPROVED** [review]
@bytrangle, I've added a few commits to resolve my comments and per the changes we discussed.

If you're good with the changes the PR is ready to be merged.

**Comment by bytrangle** [comment]
@erezrokah Thank you for your help. Do I have to do anything else?

**Comment by erezrokah** [comment]
> @erezrokah That's so quick. Thank you for your help. I'm good with your suggestions. Do I have to do anything else?

No! You've plenty and more. This is awesome 🚀

**Comment by erezrokah** [comment]
Apparently the fix in https://github.com/netlify/netlify-cms/pull/5550/commits/08e1cf340d45b43cb94a3ad1e841724806e07708#diff-31bee1c93b5344ee2a661520cb056a1ab7c7595bea9c45a1005970e222ff2718R296 revealed another issue (thanks to the tests).

It is be fixed in https://github.com/netlify/netlify-cms/pull/5550/commits/1dc5eb136038b457c9a8f08fe192ff43f39ca5ce

**Comment by bytrangle** [comment]
> Apparently the fix in [08e1cf3#diff-31bee1c93b5344ee2a661520cb056a1ab7c7595bea9c45a1005970e222ff2718R296](https://github.com/netlify/netlify-cms/commit/08e1cf340d45b43cb94a3ad1e841724806e07708#diff-31bee1c93b5344ee2a661520cb056a1ab7c7595bea9c45a1005970e222ff2718R296) revealed another issue (thanks to the tests).
> 
> It is fixed in [1dc5eb1](https://github.com/netlify/netlify-cms/commit/1dc5eb136038b457c9a8f08fe192ff43f39ca5ce)

That's so unbelievable quick. I was in the process of fixing, but it looks like I don't need to do that?

**Comment by erezrokah** [comment]
> That's so unbelievable quick. I was in the process of fixing, but it looks like I don't need to do that?

No need, thank you - the PR is merged 👍.

I'll probably release this on Sunday morning so if it breaks something for existing users the impact will be smaller (I don't expect it to break anything, but users might need to get used to the new behavior).
